src/Storages/System/StorageSystemDetachedTables.cpp

```diff
@@ -0,0 +1,236 @@
+#include <Storages/ColumnsDescription.h>
+#include <Storages/ProjectionsDescription.h>
+#include <Storages/StorageInMemoryMetadata.h>
+#include <Storages/System/StorageSystemTables.h>
+#include <Storages/System/getQueriedColumnsMaskAndHeader.h>
+#include <Storages/VirtualColumnUtils.h>
+
+#include <boost/range/adaptor/map.hpp>
+
+
+namespace DB
+{
+
+namespace
+{
+
+class DetachedTablesBlockSource : public ISource
+{
+public:
+    DetachedTablesBlockSource(
+        std::vector<UInt8> columns_mask_,
+        Block header,
+        UInt64 max_block_size_,
+        ColumnPtr databases_,
+        ColumnPtr detached_tables_,
+        ContextPtr context_)
+        : ISource(std::move(header))
+        , columns_mask(std::move(columns_mask_))
+        , max_block_size(max_block_size_)
+        , databases(std::move(databases_))
+        , context(Context::createCopy(context_))
+    {
+        size_t size = detached_tables_->size();
+        detached_tables.reserve(size);
+        for (size_t idx = 0; idx < size; ++idx)
+        {
+            detached_tables.insert(detached_tables_->getDataAt(idx).toString());
+        }
+    }
+
+    String getName() const override { return "DetachedTables"; }
+
+protected:
+    Chunk generate() override
+    {
+        if (done)
+            return {};
+
+        MutableColumns result_columns = getPort().getHeader().cloneEmptyColumns();
+
+        const auto access = context->getAccess();
+        const bool need_to_check_access_for_databases = !access->isGranted(AccessType::SHOW_TABLES);
+
+        for (size_t database_idx = 0, rows_count = 0; database_idx < databases->size() && rows_count < max_block_size; ++database_idx)
+        {
+            database_name = databases->getDataAt(database_idx).toString();
+            database = DatabaseCatalog::instance().tryGetDatabase(database_name);
+
+            if (!database)
+                continue;
+
+            const bool need_to_check_access_for_tables
+                = need_to_check_access_for_databases && !access->isGranted(AccessType::SHOW_TABLES, database_name);
+
+            if (!detached_tables_it || !detached_tables_it->isValid())
+                detached_tables_it = database->getDetachedTablesIterator(context, {}, false);
+
+            for (; rows_count < max_block_size && detached_tables_it->isValid(); detached_tables_it->next())
+            {
+                const auto detached_table_name = detached_tables_it->table();
+
+                if (!detached_tables.contains(detached_table_name))
+                    continue;
+
+                if (need_to_check_access_for_tables && !access->isGranted(AccessType::SHOW_TABLES, database_name, detached_table_name))
+                    continue;
+
+                fillResultColumnsByDetachedTableIterator(result_columns);
+                ++rows_count;
+            }
+        }
+
+        const UInt64 num_rows = result_columns.at(0)->size();
+        done = true;
```

Comment:
The previous loop can stop because we reached max rows count. In that case I think think we are not done yet.